### PR TITLE
Restructure for extensibility

### PR DESCRIPTION
**Pull request by nicklanng**
I'm reorganised Modelgen to more easily allow connecting to different database types.

The next pull request will add a command flag to select the DB (mysql, postgres, etc...) and a switch in the commands to create the correct connector and use the correct templates.

### CODE REVIEW

File: sqlfmt/common.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/LUSHDigital/modelgen/model"
+)
+
+// MapType figures out which go type should be used, based on the SQL type.
+func MapType(columnType string, nullable string, typeMap map[string]model.SQLType) (string, error) {
```
Comment (codingconcepts) [inline sqlfmt/common.go]: This is MySQL specific, so will want to go into a MySQL implementation.  (You'll need both).

File: templates/templatewriter.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	defer file.Close()
+
+	if _, err = io.Copy(file, buffer); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *TemplateWriter) WriteHelperTests() error {
```
Comment (codingconcepts) [inline templates/templatewriter.go]: This and the above function can probably be DRYed out.

File: main.go [hunk 3]
```diff
@@ -1,109 +1,19 @@
...
-	dbName   *string
-	pkgName  *string
-	conn     *string
-	database *sql.DB
-	version  string
-	box      packr.Box
+	version string
 )
 
 func init() {
```
Comment (ladydascalie) [inline main.go]: Could we get rid of this `init`?

I'd much rather we define a setup package/flow instead, so we can control it better and be consistent.

File: main.go [hunk 4]
```diff
@@ -1,109 +1,19 @@
...
 
 var (
-	output   *string
-	dbName   *string
-	pkgName  *string
-	conn     *string
-	database *sql.DB
-	version  string
-	box      packr.Box
+	version string
```
Comment (ladydascalie) [inline main.go]: If we have only one left, let's inline it into a single declaration

File: cmd/root.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	pkgName *string
+	conn    *string
+)
+
+var (
+	errFormat = errors.New("invalid connection string format")
+	rootCmd   = &cobra.Command{}
+)
+
+func init() {
```
Comment (ladydascalie) [inline cmd/root.go]: Same as before, can we get rid of the init statements?

File: cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	if conn == nil || *conn == "" {
+		log.Fatal("Please provide a connection string")
+	}
+}
+
+func deconstructConnectionString(connect string) (username, password, host, port string, err error) {
+	parts := strings.Split(connect, "@")
+	if len(parts) < 2 {
+		return "", "", "", "", errFormat
```
Comment (ladydascalie) [inline cmd/root.go]: Perhaps this could be simplified to:

```go
err = errFormat
return
```

Same thing for the ones bellow.

File: cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+func init() {
+	pkgName = rootCmd.PersistentFlags().StringP("package", "p", "generated_models", "name of package")
+	output = rootCmd.PersistentFlags().StringP("output", "o", "generated_models", "path to package")
+	dbName = rootCmd.PersistentFlags().StringP("database", "d", "", "name of database")
+	conn = rootCmd.PersistentFlags().StringP("connection", "c", "", "user:pass@host:port")
+}
+
+func Execute() {
```
Comment (ladydascalie) [inline cmd/root.go]: Missing documentation

File: connectors/mysql.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+	return &MySQL{
+		username: username,
+		password: password,
+		host:     host,
+		port:     port,
+		database: database,
+	}
+}
+
+func (m *MySQL) Connect() (*sql.DB, error) {
```
Comment (ladydascalie) [inline connectors/mysql.go]: Missing documentation. There are many other instances, it's minor but it would be a marked improvement if we could add it.

**Pull request merged**